src/System.IO.Pipelines.Networking.Tls/ApplicationProtocols.cs

```diff
@@ -0,0 +1,157 @@
+
+        [Flags]
+        public enum ProtocolIds
+        {
+            None = 0,
+            Http11 = 1,
+            Spdy1 = 2,
+            Spdy2 = 4,
+            Spdy3 = 8,
+            TraversalUsingRelaysAroundNat = 16,
+            NatDiscoveryUsingSessionTraversalUtilitiesforNat = 32,
+            Http2OverTls = 64,
+            Http2OverTcp = 128,
+            WebRtcMediaAndData = 256,
+            ConfidentialWebRtcMediaAndData = 512,
+            Ftp = 1024
+        }
+
+        internal enum ApplicaitonProtocolNegotiationExtension
+        {
+            None = 0,
+            Npn,
+            Alpn,
+        }
+
+        internal enum ApplicationProtocolNegotiationStatus : uint
+        {
+            None = 0,
+            Success,
+            SelectedClientOnly
+        }
+
+        internal static unsafe ProtocolIds FindNegotiatedProtocol(SSPIHandle context)
+        {
+            ContextApplicationProtocol protoInfo;
+            Interop.QueryContextAttributesW(ref context, ContextAttribute.ApplicationProtocol, out protoInfo);
+
+            if (protoInfo.ProtoNegoStatus != ApplicationProtocolNegotiationStatus.Success)
+            {
+                throw new InvalidOperationException("Could not negotiate a mutal application protocol");
+            }
+            return GetNegotiatedProtocol(protoInfo.ProtocolId, protoInfo.ProtocolIdSize);
+        }
+
+        public static byte[] GetBufferForProtocolId(ProtocolIds supportedProtocols, bool withAlpnHeader)
+        {
+            short listLength = 0;
+            for (int i = 0; i < _numberOfProtocols.Length; i++)
+            {
+                if (((int) _numberOfProtocols.GetValue(i) & (int) supportedProtocols) > 0)
+                {
+                    listLength += (short) (_allProtocols[i].Length + 1);
+                }
+            }
+            byte[] returnValue;
+            if (withAlpnHeader)
+            {
+                //Now we know the total length of the list lets create our array of protocols and header
+                returnValue = new byte[listLength + _HeaderLength];
+                var spa = new Span<byte>(returnValue);
+                int length = _HeaderLength + listLength - _ContentOffset;
+                spa.Write(length);
+                spa = spa.Slice(4);
+                spa.Write(ApplicaitonProtocolNegotiationExtension.Alpn);
+                spa = spa.Slice(4);
+                spa.Write(listLength);
+                spa = spa.Slice(2);
+                ProtocolList(supportedProtocols, spa);
+            }
+            else
+            {
+                //No header just make an array of the correct size
+                returnValue = new byte[listLength];
+                var spa = new Span<byte>(returnValue);
+                ProtocolList(supportedProtocols, spa);
+            }
+            return returnValue;
+        }
+
+        private static Span<byte> ProtocolList(ProtocolIds supportedProtocols, Span<byte> spa)
```

Comment:
removed the return value, no bug it writes to the span that is passed in, originally it was allocating it's own that was being appended, which obviously would anger @benaadams